src/code/su_mtx.c

```diff
@@ -0,0 +1,361 @@
+ *         Mtx_t m;
+ *         struct {
+ *             u16 intPart[4][4];
+ *             u16 fracPart[4][4];
+ *         };
+ *         long long int forc_structure_alignment;
+ *     } Mtx; // size = 0x40
+ *
+ * This means it can be written to as either words or two sets of shorts; the latter is the correct format for
+ * interpreting it the way the RSP does, but the first is likely faster to access and write to. The functions save
+ * writing halves by
+ *
+ * - writing a word 0 if two adjacent elements are 0 in a convenient place, or 1 if they are 0 and 1,
+ * - writing a word using a shift and overwriting the dirty half afterwards.
+ *
+ * Examples of both of these are seen in Mtx_SetTranslateScaleMtx(); we explain there in full detail due to the
+ * function's relative simplicity. The other two functions we merely point out the interesting parts rather than
+ * explaining the whole process.
+ *
+ * We use the following notation throughout: round brackets denote the word-sized elements, square the pairs they cover.
+ * The integer and fractional parts of a fixed-point number are denoted by i and f respectively, and these letters are
+ * also used to represent which half each of the pairs belongs to.
+ *
+ *     (00) [i00,i01], (01) [i02,i03]
+ *     (02) [i10,i11], (03) [i12,i13]
+ *     (10) [i20,i21], (11) [i22,i23]
+ *     (12) [i30,i31], (13) [i32,i33]
+ *     (20) [f00,f01], (21) [f02,f03]
+ *     (22) [f10,f11], (23) [f12,f13]
+ *     (30) [f20,f21], (31) [f22,f23]
+ *     (32) [f30,f31], (33) [f32,f33]
+ *
+ * We keep the RSP's column-major convention; this file is confusing enough without attempting to transpose everything.
+ * Bear in mind therefore that we are acting on *row vectors* on the *right*.
+ *
+ * @warning The behaviour of the output of the functions in this file is undefined in C89, since both members of the
+ * union are used to set the mtxs. (C99+ allow type-punning, but the behaviour is still implementation-defined because
+ * it relies on the storage being big-endian.)
+ *
+ * @remark Name inferred from shared Animal Forest functions, meaning of "su" is unclear.
+ */
+
+#include "global.h"
+
+/**
+ * Constructs a matrix \$f ST \$f, i.e. a scaling \$f S \$f followed by a translation \$f T \$f.
+ *
+ * The final result is
+ *
+ * \f[
+ *  \begin{bmatrix}
+ *      s_x & 0 & 0 & 0 \\
+ *      0 & s_y & 0 & 0 \\
+ *      0 & 0 & s_z & 0 \\
+ *      t_x & t_y & t_z & 1
+ *  \end{bmatrix}
+ * \f]
+ *
+ * @param[out] mtx Fixed-point matrix pointer to output to
+ * @param[in] scaleX Scale in the X direction
+ * @param[in] scaleY Scale in the Y direction
+ * @param[in] scaleZ Scale in the Z direction
+ * @param[in] translateX X component of translation
+ * @param[in] translateY Y component of translation
+ * @param[in] translateZ Z component of translation
+ *
+ * @remark Original name: "suMtxMakeTS"
+ */
+void Mtx_SetTranslateScaleMtx(Mtx* mtx, f32 scaleX, f32 scaleY, f32 scaleZ, f32 translateX, f32 translateY,
```

Comment:
Trying to think of another name for this, but most of the mathematical definitions for translate and scale also include rotate.